### PR TITLE
Rewrite multipart boundary detection

### LINKED ISSUE #7723: https://github.com/esp8266/Arduino/pull/7543 corrupt binary upload like index.html.gz
### Basic Infos

- [ X] This issue complies with the [issue POLICY doc](https://github.com/esp8266/Arduino/blob/master/POLICY.md).
- [ X] I have read the documentation at [readthedocs](https://arduino-esp8266.readthedocs.io/en/latest) and the issue is not addressed there.
- [X ] I have tested that the issue is present in current master branch (aka latest git).
- [X ] I have searched the issue tracker for a similar issue.
- [ ] If there is a stack dump, I have decoded it. N/A
- [ X] I have filled out all fields below.

#### Platform

- Hardware:					[ESP-12]
- Core Version:			[latest git 25/11/2020] 
- Development Env:		[Arduino IDE]
- Operating System:		[Windows]

### Settings in IDE

- Module:  			[Nodemcu]
- Flash Mode:			[dio]
- Flash Size: 		[4MB/FS2MB]
- lwip Variant:   [v2 Lower Memory]
- Reset Method:		[nodemcu]
- Flash Frequency:	[40Mhz]
- CPU Frequency:		[160MHz]
- Upload Using:		[SERIAL]
- Upload Speed:   [115200r] (serial upload only)

### Problem Description

Since this commit https://github.com/esp8266/Arduino/pull/7543,  binary upload like index.html.gz is corrupted
so the result is when web access to corrupted file the encodding is not recognized

Need to upload file - then try to access it using webserver
with commit the page cannot be displayed 
without commit everything is fine

No issue with text file as there is not need for browser to expand the file, even corrupted it is loaded
good file [index.html(2).gz](https://github.com/esp8266/Arduino/files/5596591/index.html.2.gz)
corrupted file [index.html(1).gz](https://github.com/esp8266/Arduino/files/5596592/index.html.1.gz)

image

For what I can see the 0D 0A code is modified in uploaded file - corrupting it

### [MCVE](https://stackoverflow.com/help/mcve) Sketch
I used FSBrowser sample to check

### PR DESCRIPTION
**Pull request by earlephilhower**
Use a simpler, cleaner implementation of multipart form detection as
defined in https://tools.ietf.org/html/rfc7578 .

Implements a simple state machine that detects the `\r\n--<boundary>`
stream in input a character at a time, instead of buffering and
comparing in chunks which can miss things due to alignment issues and
which also had a problem with replacing characters in a binary stream.

Fixes #7723

### CODE REVIEW

**Comment by earlephilhower** [comment]
@luc-github can you please test this with your failing forms and report back?

Using FSBrowser and your binaries, it works correctly, but I don't have any other handy multipart form apps to test against.

**Comment by luc-github** [comment]
ok I will test when back home and report results

thank you

**Comment by luc-github** [comment]
I can confirm the PR fix the issue - I have tested with my application ESP3D  and now index.html.gz is properly expanded after upload - which confirm no more corruption during upload

Thank you for the fix

File: libraries/ESP8266WebServer/src/Parsing-impl.h [hunk 1]
```diff
@@ -444,45 +444,33 @@ bool ESP8266WebServerTemplate<ServerType>::_parseForm(ClientType& client, const
...
-                if (first) first = false;   //only add newline characters after the first line
-                else {
-                    _uploadWriteByte('\r');
-                    _uploadWriteByte('\n');
+            int fastBoundaryLen = 4 /* \r\n-- */ + boundary.length() + 1 /* \0 */;
+            char fastBoundary[ fastBoundaryLen ];
+            snprintf(fastBoundary, fastBoundaryLen, "\r\n--%s", boundary.c_str());
+            int boundaryPtr = 0;
+            while ( true ) {
+                if ( !client.connected() ) {
```
Comment (d-a-v) [inline libraries/ESP8266WebServer/src/Parsing-impl.h:452]: We could be tolerant with premature peer closing when it sends all data anyway.
With `if ( !client.available() && !client.connected() )`, bailing out is conditioned by:
- peer disconnection, *and*
- receive buffer is fully read

(same kind of test is made if `_uploadReadByte()`)

Comment (earlephilhower) [reply libraries/ESP8266WebServer/src/Parsing-impl.h]: Done by refactoring the uploadReadByte routine to not throw away failure.

**Review by d-a-v — APPROVED** [review]

**Pull request merged**